libsolidity/interface/CompilerStack.cpp

```diff
@@ -1206,13 +1206,93 @@ Json CompilerStack::ethdebug(Contract const& _contract, bool _runtime) const
 	solAssert(m_stackState >= AnalysisSuccessful, "Analysis was not successful.");
 	solAssert(_contract.contract);
 	solUnimplementedAssert(!isExperimentalSolidity());
-	if (_runtime)
+	Json result = {
+		{
+			"contract",
+			{
+					{"name", _contract.contract->name()},
+					{
+						"definition",
+						{{
+							"sources",
+							{{"id", static_cast<int>(sourceIndices()[_contract.contract->sourceUnitName()])}}
+						}}
+					}
+			}
+		},
+		{"environment ", _runtime ? "call" : "create"},
+		{"instructions", ethdebugInstructions(_contract, _runtime)}
+	};
+	return result;
+}
+
+Json CompilerStack::ethdebugInstructions(Contract const& _contract, bool _runtime) const
+{
+	evmasm::LinkerObject const& object = _runtime ? _contract.runtimeObject : _contract.object;
+	std::shared_ptr<evmasm::Assembly> const& assembly = _runtime ? _contract.evmRuntimeAssembly : _contract.evmAssembly;
+
+	// e.g. interfaces don't have a valid assembly object.
+	if (assembly)
 	{
-		Json result = Json::object();
-		return result;
+		solUnimplementedAssert(assembly->eofVersion() == std::nullopt, "ethdebug does not yet support EOF.");
+		solUnimplementedAssert(assembly->codeSections().size() == 1, "ethdebug does not yet support multiple code-sections.");
 	}
-	Json result = Json::object();
-	return result;
+
+	Json instructions = Json::array();
+	for (size_t i = 0; i < object.bytecodeOffsets.size(); ++i)
+	{
+		size_t offset = object.bytecodeOffsets[i];
+		size_t nextOffset = offset;
+		if (i + 1 < object.bytecodeOffsets.size())
+			nextOffset = object.bytecodeOffsets[i + 1];
+		solAssert(offset < object.bytecode.size());
+		solAssert(nextOffset < object.bytecode.size());
+		Json operation{{
+			"mnemonic", instructionInfo(static_cast<evmasm::Instruction>(object.bytecode[offset]), m_evmVersion).name
+		}};
+		if (offset != nextOffset)
+		{
+			static size_t constexpr instructionSize = 1;
+			bytes const argumentData(
+				object.bytecode.begin() + static_cast<std::ptrdiff_t>(offset) + instructionSize,
+				object.bytecode.begin() + static_cast<std::ptrdiff_t>(nextOffset)
+			);
```

Comment:
This relies on some strong assumptions: that offsets are set for all instructions, that there's nothing else between them and that they're in ascending order. Otherwise you will grab too much stuff into arguments.

I don't think this is true in general. Take for example `AssignImmutable`, which translates to a variable number of `POP`/`DUP2`/`MSTORE`/`ADD` instructions, but you only store one offset for the whole assembly item.

And while the order should be fine the way it's implemented now, `LinkerObject` is a struct with free access to its fields, there are no sanity checks and the assumptions are not even documented. It would be very easy to break them accidentally later, e.g. when Radek will be implementing this for EOF.

Reply:
hmm.. oh yes, good point!